crypto/rsa/rsa_sign.c

```diff
@@ -16,60 +16,189 @@
+{
+    switch (nid) {
+#ifndef FIPS_MODE
+    MD_CASE_NN(sha1);
+#endif
+    MD_CASE_NN(sha224);
+    MD_CASE_NN(sha256);
+    MD_CASE_NN(sha384);
+    MD_CASE_NN(sha512);
+    MD_CASE_NN(sha512_224);
+    MD_CASE_NN(sha512_256);
+    MD_CASE_NN(sha3_224);
+    MD_CASE_NN(sha3_256);
+    MD_CASE_NN(sha3_384);
+    MD_CASE_NN(sha3_512);
+    default:
+        return NULL;
+    }
+}
+
 /* Size of an SSL signature: MD5+SHA1 */
 #define SSL_SIG_LENGTH  36
 
 /*
  * encode_pkcs1 encodes a DigestInfo prefix of hash |type| and digest |m|, as
  * described in EMSA-PKCS1-v1_5-ENCODE, RFC 3447 section 9.2 step 2. This
  * encodes the DigestInfo (T and tLen) but does not add the padding.
+ * 
+ * When verifying, decrypt_len indicates the size of the decrypted DigestInfo
+ * and is used to determine if the alternative encoding should be used for
+ * comparison.  When signing, decrypt_len is zero.
  *
  * On success, it returns one and sets |*out| to a newly allocated buffer
  * containing the result and |*out_len| to its length. The caller must free
  * |*out| with |OPENSSL_free|. Otherwise, it returns zero.
  */
-static int encode_pkcs1(unsigned char **out, int *out_len, int type,
-                        const unsigned char *m, unsigned int m_len)
+static int encode_pkcs1(unsigned char **out, size_t *out_len,
+                        size_t decrypt_len, int type,
+                        const unsigned char *m, size_t m_len)
 {
-    X509_SIG sig;
-    X509_ALGOR algor;
-    ASN1_TYPE parameter;
-    ASN1_OCTET_STRING digest;
-    uint8_t *der = NULL;
-    int len;
-
-    sig.algor = &algor;
-    sig.algor->algorithm = OBJ_nid2obj(type);
-    if (sig.algor->algorithm == NULL) {
+    size_t di_prefix_len, dig_info_len;
+    const unsigned char *di_prefix;
+    unsigned char *dig_info;
+
+    if (type == NID_undef) {
         RSAerr(RSA_F_ENCODE_PKCS1, RSA_R_UNKNOWN_ALGORITHM_TYPE);
         return 0;
     }
-    if (OBJ_length(sig.algor->algorithm) == 0) {
+    di_prefix = digestinfo_encoding(type, &di_prefix_len);
+    if (di_prefix == NULL) {
         RSAerr(RSA_F_ENCODE_PKCS1,
                RSA_R_THE_ASN1_OBJECT_IDENTIFIER_IS_NOT_KNOWN_FOR_THIS_MD);
         return 0;
     }
-    parameter.type = V_ASN1_NULL;
-    parameter.value.ptr = NULL;
-    sig.algor->parameter = &parameter;
-
-    sig.digest = &digest;
-    sig.digest->data = (unsigned char *)m;
-    sig.digest->length = m_len;
+    dig_info_len = di_prefix_len + m_len;
+    if (decrypt_len != 0 && dig_info_len != decrypt_len) {
+        /* Use alternate encoding (parameters not present) */
+        size_t enc_len;
+        const unsigned char *enc;
+        
+        if ((enc = digestinfo_encoding_nn(type, &enc_len)) != NULL) {
+            di_prefix = enc;
+            di_prefix_len = enc_len;
+            dig_info_len = di_prefix_len + m_len;
+        }
+    }
 
-    len = i2d_X509_SIG(&sig, &der);
-    if (len < 0)
+    dig_info = OPENSSL_malloc(dig_info_len);
```

Comment:
why is ```enc``` assignment at line 177 inside the if, and ```dig_info``` here is before it? consistency please.